sgkit/io/vcf/utils.py

```diff
@@ -19,16 +19,23 @@ def ceildiv(a: int, b: int) -> int:
     return -(-a // b)
 
 
-# https://dev.to/orenovadia/solution-chunked-iterator-python-riddle-3ple
+# Based on https://dev.to/orenovadia/solution-chunked-iterator-python-riddle-3ple
 def chunks(iterator: Iterator[T], n: int) -> Iterator[Iterator[T]]:
```

Comment:
Nice.

For future I dig up some implementation I did some time again, in case there's any interest in that:

```
from itertools import chain, islice

import six
import typing


def batch(iterable: typing.Iterable, size: int) -> typing.Generator: 
    """
    For example:
    ```
    list(batch([1, 2, 3], 2)) -> [[1, 2], [3]]
    ```
    """
    if size <= 0:
        raise ValueError("Can't iterate over batch size <= 0")
    if iterable is None:
        raise ValueError("Iterable can't be None")
    if hasattr(iterable, "__len__"):
        # leverage len to do efficient batching
        length = len(iterable)  # type: ignore
        it = iter(iterable)
        for i in range(0, length, size):
            yield [k for k in islice(it, size)]
    else:
        # fallback to generator based method
        for chunk in batch_gen(iterable, size):
            yield chunk
        return


def batch_gen(generator: typing.Iterable, size: int, lazy: bool=False) -> typing.Generator:
    """
    For example:
    ```
    def foo_gen():
        for x in range(5):
            yield x

    list(batch(foo_gen(), 2)) -> [[0, 1], [2, 3], [4]]
    ```

    NOTE: default implementation is eager on the materialization of the elements, you can use lazy parameter to
    get batches of generators of elements, but keep in mind that lazy implementation is probably not thread safe, and
    batch element generator has to be exhausted before retrieving the next batch.
    """
    if size <= 0:
        raise ValueError("Can't iterate over batch size <= 0")
    if generator is None:
        raise ValueError("generator can't be None")
    it = iter(generator)
    while True:
        batch_it = islice(it, size)
        try:
            chunk_gen = chain([six.next(batch_it)], batch_it)
        except StopIteration:
            return
        if lazy:
            yield chunk_gen
        else:
            yield [x for x in chunk_gen]
```

Reply:
Thanks @ravwojdyla! Is the batch code more efficient do you think? Something worth evaluating in the future perhaps.

Reply:
@tomwhite afair, it might be if the iterable has length (`hasattr(iterable, "__len__")`).

Reply:
Ah, we don't know the number of variants in a particular region, so I don't think we can use this optimization unfortunately.